### PR TITLE
Fix some utility functions

### PR DESCRIPTION
**Pull request by stramel**
- Ignores `package-lock.json` and disables `package-lock.json` from being generated
- Adds: `isBoolean`, `isFunction`
- Adds tests for utility functions
- Adds typings for ReactDOM

### CODE REVIEW

**Comment by bluebill1049** [comment]
wow!

File: src/logic/getValueAndMessage.ts [hunk 1]
```diff
@@ -1,15 +1,13 @@
 import isNullOrUndefined from '../utils/isNullOrUndefined';
 import { DataType } from '../types';
+import isObject from '../utils/isObject';
 
 export default (
-  item: DataType | number | string | undefined,
+  item?: DataType | number | string,
```
Comment (bluebill1049) [inline src/logic/getValueAndMessage.ts]: 👍

File: src/logic/getValueAndMessage.ts [hunk 2]
```diff
@@ -1,15 +1,13 @@
...
   value: number | string | RegExp;
   message: string;
 } => ({
-  value:
-    typeof item === 'object' && !isNullOrUndefined(item.value)
-      ? item.value
-      : item,
-  message: typeof item === 'object' && item.message ? item.message : '',
+  value: isObject(item) && !isNullOrUndefined(item.value) ? item.value : item,
+  message: isObject(item) && item.message ? item.message : '',
```
Comment (bluebill1049) [inline src/logic/getValueAndMessage.ts]: 👍

File: src/logic/validateField.ts [hunk 3]
```diff
@@ -131,12 +133,9 @@ export default async <Data>(
     const fieldValue = isRadio ? getRadioValue(options).value : value;
     const validateRef = isRadio && options ? options[0].ref : ref;
 
-    if (typeof validate === 'function') {
+    if (isFunction(validate)) {
```
Comment (bluebill1049) [inline src/logic/validateField.ts]: 👍

File: src/utils/flatArray.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 export default function flatArray<T>(list: T[]): T[] {
-  return list.reduce(
-    (a: any, b: any) => a.concat(Array.isArray(b) ? flatArray(b) : b),
+  return list.reduce<T[]>(
```
Comment (bluebill1049) [inline src/utils/flatArray.ts]: 😮

**Comment by stramel** [comment]
😅 I'm working on a bigger overall type rework. This was just a small change that  I think can easily go in and get benefits from.

File: src/utils/isBoolean.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export default (value: unknown): value is boolean => typeof value === 'boolean';
```
Comment (bluebill1049) [inline src/utils/isBoolean.ts]: great use of `unknown`

File: src/utils/isUndefined.ts [hunk 6]
```diff
@@ -1 +1 @@
-export default (val: any) => val === undefined;
+export default (val: unknown): val is undefined => val === undefined;
```
Comment (bluebill1049) [inline src/utils/isUndefined.ts]: 😮

**Review by bluebill1049 — APPROVED** [review]
Thank you @stramel so good!

**Comment by bluebill1049** [comment]
merge whenever you are ready :)

**Comment by stramel** [comment]
@bluebill1049 	`isObject` has various implementations across util libraries. Which of these are you intending for it to return true on `[]`, `{}`, `Function`?

**Comment by bluebill1049** [comment]
@stramel I think we want 
```
isObject {}
isFunction {}
isArray []
```

**Comment by barrymay** [comment]
nice :)

**Pull request merged**